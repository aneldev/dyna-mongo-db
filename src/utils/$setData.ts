import {dynaObjectScan} from "dyna-object-scan";

export const $setData = (data: any, propertyName?: string): { [p: string]: any } => {
  const output = {};
  dynaObjectScan(data, ({path, value}) => {
    if (value === undefined) return;
    const applyPropertyName =
      propertyName
        ? propertyName + path
        : (path || '').substr(1);
    output[applyPropertyName] = value;
  });

  return output;
};
