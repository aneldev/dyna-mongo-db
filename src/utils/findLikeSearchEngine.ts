export const findLikeSearchEngine = (fieldName: string, searchText?: string) => {
  if (!searchText || !searchText.trim()) return {};
  const searchParts = searchText.split(' ').filter(Boolean);
  return {
    $and: searchParts
      .map(searchPart => {
        const isNot = searchPart[0] === '-';
        const text = isNot ? searchPart.substr(1) : searchPart;
        const comparison = {
          $regex: `.*${text}.*`,
          $options: 'i',
        };
        return isNot ?
          {[fieldName]: {$not: comparison}}
          : {[fieldName]: comparison};
      }),
  };
};
