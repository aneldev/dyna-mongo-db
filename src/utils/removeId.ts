export const removeId = <TData, >(data: TData): Omit<TData, "id"> => {
  const data_: any = {...data};
  delete data_._id;
  return data_;
};
