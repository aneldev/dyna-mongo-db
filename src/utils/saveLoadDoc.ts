import {ObjectId} from "mongodb";

export const saveDoc = <TData, >(item: TData): any => {
  const item_: any = {...item};
  if (item_.id !== undefined) {
    if (item_.id) item_._id = new ObjectId(item_.id);
    delete item_.id;
  }
  return item_;
};

export const loadDoc = <TData, >(item: TData): TData => {
  const item_: any = {...item};
  item_.id = item_._id.toHexString();
  delete item_._id;
  return item_;
};
