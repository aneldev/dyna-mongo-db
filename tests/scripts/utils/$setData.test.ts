import {$setData} from "../../../src/utils/$setData";

describe('$setData', () => {
  test('Plain object', () => {
    expect(
      $setData({
        name: 'John',
        phones: {
          main: '+44000011112222',
        },
      }, 'user'),
    )
      .toMatchSnapshot();
  });
  test('Root property', () => {
    expect(
      $setData({
        name: 'John',
        phones: {
          main: '+44000011112222',
        },
      }),
    )
      .toMatchSnapshot();
  });
  test('Null and undefined values', () => {
    expect(
      $setData({
        firstName: 'John',
        middleName: null,
        lastName: undefined,
        phones: {
          main: '+44000011112222',
        },
      }, 'user'),
    )
      .toMatchSnapshot();
  });
  test('Data with array', () => {
    expect(
      $setData({
        name: 'John',
        cars: [
          {brand: 'Volvo', modal: 2920},
          {brand: 'Saab', modal: 2027},
        ],
      }, 'user'),
    )
      .toMatchSnapshot();
  });
});
