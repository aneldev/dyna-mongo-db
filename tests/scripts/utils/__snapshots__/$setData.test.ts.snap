// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$setData Data with array 1`] = `
Object {
  "user.cars": Array [
    Object {
      "brand": "Volvo",
      "modal": 2920,
    },
    Object {
      "brand": "Saab",
      "modal": 2027,
    },
  ],
  "user.cars[0]": Object {
    "brand": "Volvo",
    "modal": 2920,
  },
  "user.cars[0].brand": "Volvo",
  "user.cars[0].modal": 2920,
  "user.cars[1]": Object {
    "brand": "Saab",
    "modal": 2027,
  },
  "user.cars[1].brand": "Saab",
  "user.cars[1].modal": 2027,
  "user.name": "John",
}
`;

exports[`$setData Null and undefined values 1`] = `
Object {
  "user.firstName": "John",
  "user.middleName": null,
  "user.phones": Object {
    "main": "+44000011112222",
  },
  "user.phones.main": "+44000011112222",
}
`;

exports[`$setData Plain object 1`] = `
Object {
  "user.name": "John",
  "user.phones": Object {
    "main": "+44000011112222",
  },
  "user.phones.main": "+44000011112222",
}
`;

exports[`$setData Root property 1`] = `
Object {
  "name": "John",
  "phones": Object {
    "main": "+44000011112222",
  },
  "phones.main": "+44000011112222",
}
`;
